### ISSUE TITLE
datarace jb.completed on ^C

### ISSUE THREAD

**Issue by mb0** [mheese/journalbeat#98]
data race occurred with release version 5.5 for debian, but most likely applies to other builds of that version.

this data races is not clear to me. but as a rule of thumb only a single write go routine should ever close
a channel if not strictly used for signaling. if not coordinating channel close and write one will run into panics because of a send on a close channel.

the completed channel is used for both (for range in managePendingQueueLoop defer). normally i would propose yet another channel to split up one for messages and one to broadcast a close. but in this situation i would maybe try to move the defer flush in managePendingQueueLoop to the defer in Run and coordinate Run to wait for managePendingQueueLoop before cleanup. or any other change that simplifies the the shutdown procedure.
```
==================
WARNING: DATA RACE
Write at 0x00c420117740 by main goroutine:
  runtime.closechan()
      /root/.gimme/versions/go1.8.linux.amd64/src/runtime/chan.go:320 +0x0
  github.com/mheese/journalbeat/beater.(*Journalbeat).Run.func1()
      /go/src/github.com/mheese/journalbeat/beater/journalbeat.go:177 +0xd6
  github.com/mheese/journalbeat/beater.(*Journalbeat).Run()
      /go/src/github.com/mheese/journalbeat/beater/journalbeat.go:219 +0xb31
  github.com/mheese/journalbeat/vendor/github.com/elastic/beats/libbeat/beat.(*Beat).launch()
      /go/src/github.com/mheese/journalbeat/vendor/github.com/elastic/beats/libbeat/beat/beat.go:225 +0xa5f
  github.com/mheese/journalbeat/vendor/github.com/elastic/beats/libbeat/beat.Run()
      /go/src/github.com/mheese/journalbeat/vendor/github.com/elastic/beats/libbeat/beat/beat.go:141 +0x72
  main.main()
      /go/src/github.com/mheese/journalbeat/main.go:25 +0x65

Previous read at 0x00c420117740 by goroutine 19:
  runtime.chansend()
      /root/.gimme/versions/go1.8.linux.amd64/src/runtime/chan.go:128 +0x0
  github.com/mheese/journalbeat/beater.(*eventSignal).Completed()
      /go/src/github.com/mheese/journalbeat/beater/state.go:43 +0x5f
  github.com/mheese/journalbeat/vendor/github.com/elastic/beats/libbeat/common/op.(*cancelableSignal).Completed()
      /go/src/github.com/mheese/journalbeat/vendor/github.com/elastic/beats/libbeat/common/op/signal.go:182 +0xdc
  github.com/mheese/journalbeat/vendor/github.com/elastic/beats/libbeat/common/op.(*compositeSignal).Completed()
      /go/src/github.com/mheese/journalbeat/vendor/github.com/elastic/beats/libbeat/common/op/signal.go:146 +0xb8
  github.com/mheese/journalbeat/vendor/github.com/elastic/beats/libbeat/common/op.SigCompleted()
      /go/src/github.com/mheese/journalbeat/vendor/github.com/elastic/beats/libbeat/common/op/signal_util.go:18 +0x52
  github.com/mheese/journalbeat/vendor/github.com/elastic/beats/libbeat/outputs/mode/single.(*Mode).publish()
      /go/src/github.com/mheese/journalbeat/vendor/github.com/elastic/beats/libbeat/outputs/mode/single/single.go:152 +0x4e6
  github.com/mheese/journalbeat/vendor/github.com/elastic/beats/libbeat/outputs/mode/single.(*Mode).PublishEvents()
      /go/src/github.com/mheese/journalbeat/vendor/github.com/elastic/beats/libbeat/outputs/mode/single/single.go:99 +0xcc
  github.com/mheese/journalbeat/vendor/github.com/elastic/beats/libbeat/outputs/elasticsearch.(*elasticsearchOutput).BulkPublish()
      /go/src/github.com/mheese/journalbeat/vendor/github.com/elastic/beats/libbeat/outputs/elasticsearch/output.go:415 +0x9e
  github.com/mheese/journalbeat/vendor/github.com/elastic/beats/libbeat/publisher.(*outputWorker).sendBulk()
      /go/src/github.com/mheese/journalbeat/vendor/github.com/elastic/beats/libbeat/publisher/output.go:112 +0x1ca
  github.com/mheese/journalbeat/vendor/github.com/elastic/beats/libbeat/publisher.(*outputWorker).onBulk()
      /go/src/github.com/mheese/journalbeat/vendor/github.com/elastic/beats/libbeat/publisher/output.go:88 +0xbe
  github.com/mheese/journalbeat/vendor/github.com/elastic/beats/libbeat/publisher.(*outputWorker).onMessage()
      /go/src/github.com/mheese/journalbeat/vendor/github.com/elastic/beats/libbeat/publisher/output.go:70 +0xac
  github.com/mheese/journalbeat/vendor/github.com/elastic/beats/libbeat/publisher.(*messageWorker).onEvent()
      /go/src/github.com/mheese/journalbeat/vendor/github.com/elastic/beats/libbeat/publisher/worker.go:83 +0xa7
  github.com/mheese/journalbeat/vendor/github.com/elastic/beats/libbeat/publisher.(*messageWorker).run()
      /go/src/github.com/mheese/journalbeat/vendor/github.com/elastic/beats/libbeat/publisher/worker.go:69 +0x2f6

Goroutine 19 (running) created at:
  github.com/mheese/journalbeat/vendor/github.com/elastic/beats/libbeat/publisher.(*messageWorker).init()
      /go/src/github.com/mheese/journalbeat/vendor/github.com/elastic/beats/libbeat/publisher/worker.go:57 +0x17c
  github.com/mheese/journalbeat/vendor/github.com/elastic/beats/libbeat/publisher.newOutputWorker()
      /go/src/github.com/mheese/journalbeat/vendor/github.com/elastic/beats/libbeat/publisher/output.go:55 +0x324
  github.com/mheese/journalbeat/vendor/github.com/elastic/beats/libbeat/publisher.(*BeatPublisher).init()
      /go/src/github.com/mheese/journalbeat/vendor/github.com/elastic/beats/libbeat/publisher/publish.go:246 +0x127a
  github.com/mheese/journalbeat/vendor/github.com/elastic/beats/libbeat/publisher.New()
      /go/src/github.com/mheese/journalbeat/vendor/github.com/elastic/beats/libbeat/publisher/publish.go:197 +0x156
  github.com/mheese/journalbeat/vendor/github.com/elastic/beats/libbeat/beat.(*Beat).launch()
      /go/src/github.com/mheese/journalbeat/vendor/github.com/elastic/beats/libbeat/beat/beat.go:193 +0x5d9
  github.com/mheese/journalbeat/vendor/github.com/elastic/beats/libbeat/beat.Run()
      /go/src/github.com/mheese/journalbeat/vendor/github.com/elastic/beats/libbeat/beat/beat.go:141 +0x72
  main.main()
      /go/src/github.com/mheese/journalbeat/main.go:25 +0x65
==================
```

**Comment by nicorevin** [comment]
Hi @mb0! Thanks for reporting. I know this bug unfortunately...

The problem is that the completed channel is used for collecting event references from the publisher. Currently it works as follows:
1. you publish a message and attach its reference to the Signler.
2. Once the message is acked the ref is sent to the completed channel.
3. When you ^C you send a stop signal to the publisher and it cancels all in-flight events and stops working and _in theory_ this all should along with recieving last event acknowledgements should happen before the completed queue is closed.
I tried to address this issue here https://github.com/mheese/journalbeat/commit/f0fc9864a17acaf53e548372959d34e1e1171a6e making the completed queue buffered.
And I think I've addressed the #97 as well. I will re-check shortly.

**Comment by nicorevin** [comment]
@mb0 if you could test out https://github.com/mheese/journalbeat/commit/f0fc9864a17acaf53e548372959d34e1e1171a6e I would highly appreciate. My tests show that both data races you reported seem to be gone.

**Comment by mb0** [comment]
both data races still exist on f0fc986. the details i described have not changed. should i fix it send a patch?

**Comment by mb0** [comment]
with the simple change below both races disappear. i am not sure why the race on complete is also gone, but that is probably your fix working as intended.

there is still a race in libbeat/logp but that is already fixed upstream. should i open another issue to update the dependency?

```
diff --git a/beater/journalbeat.go b/beater/journalbeat.go
index 04e782e..951985e 100644
--- a/beater/journalbeat.go
+++ b/beater/journalbeat.go
@@ -184,6 +184,8 @@ func New(b *beat.Beat, cfg *common.Config) (beat.Beater, error) {
 func (jb *Journalbeat) Run(b *beat.Beat) error {
        logp.Info("Journalbeat is running!")
        defer func() {
+               _ = jb.client.Close()
+               close(jb.completed)
                _ = jb.journal.Close()
                close(jb.cursorChan)
                close(jb.pending)
@@ -235,6 +237,4 @@ func (jb *Journalbeat) Run(b *beat.Beat) error {
 func (jb *Journalbeat) Stop() {
        logp.Info("Stopping Journalbeat")
        close(jb.done)
-       _ = jb.client.Close()
-       close(jb.completed)
 }
```

**Comment by mb0** [comment]
the commit that fixed the logp data race is elastic/beats@4e34814e

**Comment by nicorevin** [comment]
Nice catch!

Once the `jb.wg.Wait()` was added everything but `close(jb.done)` should be removed from the `(jb *Journalbeat) Stop()`

Please open up a PR.

as for the logp data race make it a separate PR please.

**Comment by nicorevin** [comment]
the race is gone because now everything is done in one goroutine.

**Comment by mb0** [comment]
ha, i did not notice the sends were from that go routine, should have been obvious from the race output.
